### PR TITLE
Add chapter headings to article pages 

### PR DESCRIPTION
**Pull request by holder-josh**
w/ @jennysivapalan 

This feature is behind a feature switch in Composer, so for now this only current affects one article: http://www.theguardian.com/travel/2015/jun/06/isles-of-scilly-holiday-guide

screen shot 2015-09-07 at 17 28 22

It's also behind a switch in Frontend.

(Styled after the embed version of this which has been floating around for months: http://www.theguardian.com/travel/2015/aug/24/lyon-france-alternative-city-guide)

### CODE REVIEW

File: static/src/stylesheets/content.scss [hunk 1]
```diff
@@ -53,6 +53,7 @@
 @import 'module/content/_gallery';
 @import 'module/content/_gallery-lightbox';
 @import 'module/content/_block-level-sharing';
+@import 'module/_chapters';
```
Comment (OliverJAsh) [inline static/src/stylesheets/content.scss]: These should probably be included in the inline CSS so we don't get any FOUC.

File: common/app/views/support/HtmlCleaner.scala [hunk 2]
```diff
@@ -602,3 +603,24 @@ object MembershipEventCleaner extends HtmlCleaner {
...
+      .trim.replaceAll(" +", "-")
+  }
+
+  override def clean(document: Document): Document = {
+    if(ChapterHeadingsSwitch.isSwitchedOn) {
+      val autoaChapters = document.getElementsByClass("auto-chapter")
+
+      autoaChapters.foreach { ch =>
+        val h2 = ch.getElementsByTag("h2")
+        h2.attr("id", slugify(h2.text())).attr("class", "anchor-link-fix")
```
Comment (OliverJAsh) [inline common/app/views/support/HtmlCleaner.scala]: Maybe we can target elements with the sticky nav fix a different way. Will have a think…

Comment (OliverJAsh) [inline common/app/views/support/HtmlCleaner.scala]: We could apply the fix styles to all `h2`s? See here for more info on the hack: http://stackoverflow.com/questions/11386807/adding-an-automatic-offset-to-the-scroll-position-for-all-hash-links-calls

Comment (Jholder112233) [inline common/app/views/support/HtmlCleaner.scala]: @OliverJAsh it wouldn't even need to be all H2s, just H2s within sections with the class "auto-chapter". 

I did it this way though so this fix could be applied to other anchor link elements. e.g. the comment box at the bottom of the page.

File: static/src/stylesheets/base/_links.scss [hunk 3]
```diff
@@ -21,3 +21,8 @@ a[name] {
     display: inline-block;
     width: 0px;
 }
+
+.anchor-link-fix {
+    padding-top: $gs-baseline*4.5;
+    margin-top: $gs-baseline*-4.5;
+}
```
Comment (OliverJAsh) [inline static/src/stylesheets/base/_links.scss]: Perhaps document what the hack is (Stack Overflow link) and why it is needed (sticky nav)

File: common/app/views/support/HtmlCleaner.scala (hunk 2 above)
Comment (OliverJAsh) [inline common/app/views/support/HtmlCleaner.scala]: Ah, I see! Can we only add this attribute if the overloaded `Viewability` switch is enabled, so we remember to remove it if we decide to remove the sticky nav. Might also be worth documenting why it is needed here too (sticky nav).

File: static/src/stylesheets/base/_links.scss [hunk 4]
```diff
@@ -21,3 +21,10 @@ a[name] {
     display: inline-block;
```
Comment (Jholder112233) [inline static/src/stylesheets/base/_links.scss]: Hey @Calanthe is this hack in use currently? I had a quick look and couldn't find where this would apply

**Comment by Jholder112233** [comment]
@OliverJAsh fancy giving this a +1? Sam is happy for it to go from a design pov to test functionality.

**Comment by OliverJAsh** [comment]
:+1: Nice!

**Pull request merged**